### DISCUSSION TITLE
Pointers for running this on a GPU via a cloud service?

### DISCUSSION THREAD

**Discussion by jaypinho** [openai/whisper#398] [Q&A] [answered]
I have an M1 MacBook Pro and, due to known issues (discussed elsewhere, including #91), Whisper doesn't make use of the GPU, so transcription is rather slow.

So as an alternative I'm interested in running Whisper in the cloud using a major provider's (AWS, Azure, GCP, etc.) GPU services.

I'm a relatively comfortable user of AWS (Lambda especially), but have no experience running software that specifically needs to run on a GPU. Are there any tutorials or blog posts, etc. that would specifically apply to Whisper that I could follow to do this? Thank you!

**Comment by aadnk** [comment, marked as answer]
I wrote a detailed guide for getting my [WebUI](https://huggingface.co/spaces/aadnk/whisper-webui) to run on Google Colab for free, but you can use whatever WebUI you like (or even just run Whisper directly via the CLI/Python):

* [Reddit Tutorial](https://www.reddit.com/r/LearnJapanese/comments/xljc3e/whisper_a_new_free_ai_model_from_openai_that_can/iplegg8/)

Just be careful to disconnect/deallocate the GPU instance once you're done, so that you don't end up using all your free GPU credits, though you can optionally buy more and also get access to a more powerful GPU. And course, you can always download the Jupyter Notebook itself and upload it to an alternative provider, if they happen to be cheaper than Google. 

I've also added support [for Docker](https://huggingface.co/spaces/aadnk/whisper-webui/blob/main/dockerfile) in the WebUI, if you can find a provider that supports it and is easy to set up. The Docker container is currently [hosted on GitLab](https://gitlab.com/aadnk/whisper-webui/container_registry/3483634).

**Reply by jaypinho** [reply]
Thank you, @aadnk ! This is incredibly helpful.

**Reply by kalevleetaru** [reply]
Here's my step-by-step for GCP's GCE, including precise benchmarks, tradeoffs of T4 vs V100, etc:

https://github.com/openai/whisper/discussions/395

It includes the entire installation process, including a few extra libraries needed for the GCP "deep learning" VM image and how to reinstall the driver on reboot as needed. Its pretty straightforward to install on a cloud VM and run and you should be up and running in just a few minutes.

**Reply by jaypinho** [reply]
Thank you @kalevleetaru ! Will check it out.

**Comment by jkseppan** [comment]
Here's how I have run Whisper on an AWS EC2 GPU instance at a pretty low cost:

- start an instance through the AWS console; I have used `Deep Learning AMI GPU PyTorch 1.12.? (Amazon Linux 2)` as the AMI and `p3.2xlarge` as the instance type (that one has one GPU and costs a little over $3/hour)
- connect to the instance with ssh, run the suggested conda command to activate pytorch
- run `pip install git+https://github.com/openai/whisper.git`
- now you have a `whisper` command available - the first run takes a while, so be patient
- remember to terminate the instance when you are done

You can save money by using spot instances, but at least in my account the initial quota for p3 spot instances was zero. You have to request an increase from AWS support; a `p3.2xlarge` has 8 vCPUs so ask for a budget of 8.

**Reply by sureronald** [reply]
how was the performance (benchmarks if any)?

**Comment by louismorgner** [comment]
Just for future reference: we are now building on replicate.com with serverless GPUs that make our lives a lot easier for hosting whisper in production.

**Comment by crestrepoq** [comment]
There is a way to use it with Lambda? I have been trying whit Lambda Container Image Triggered by S3 when is uploaded a file, but I have had some troubles.

This is the docker file I'm building:
```
FROM public.ecr.aws/lambda/python:3.9

FROM katalonstudio/katalon
RUN apt-get -y update &&  apt-get install -y ffmpeg 
RUN apt-get update && apt-get install -y python3-pip

RUN mkdir /tmp/data

ENV TRANSFORMERS_CACHE "/tmp/data"

COPY requirements.txt .
RUN pip3 install -r requirements.txt
COPY app.py .

CMD [ "python3", "app.py" ]
```
This is the app.py for test:
```
import os
import json
import urllib.parse
import whisper
import boto3

s3 = boto3.client("s3")

def handler(event, context):
    try:
        bucket = event["Records"][0]["s3"]["bucket"]["name"]
        key = event['Records'][0]['s3']['object']['key']
        print("Bucket:", bucket, "key:", key)
        
        # Definir la nueva carpeta /tmp/data
        os.makedirs("/tmp/data", exist_ok=True)

        audio_file = '/tmp/data/{}'.format(key)
        print(audio_file)
        print(os.listdir()) # Chekcing files before download audio

        # Downloading file to transcribe
        s3.download_file(bucket, key, audio_file)
        print(os.listdir()) # Chekcing if file was downloaded

        # Configuring a variable needed by Huggingface I do not know why
        os.environ["TRANSFORMERS_CACHE"] = "/tmp/data"

        model = whisper.load_model("base")
        print("Model is up")
        result = model.transcribe(audio_file, fp16=False)
        print(os.listdir())
        return {
            "statusCode": 200,
            "body": json.dumps(result["text"])
        }
    except Exception as e:
        print(e)
        return {
            "statusCode": 500,
            "body": json.dumps("Error processing the file")
        }
```

This is the requirements.txt:
```
boto3
openai-whisper
setuptools-rust
```

This is the current error:
```
+ echo Entrypoint
Entrypoint
+ '[' -z '' ']'
+ exec python3 app.py
+ echo Entrypoint
Entrypoint
+ '[' -z '' ']'
+ exec python3 app.py
OpenBLAS WARNING - could not determine the L2 cache size on this system, assuming 256k
RequestId: e06adc5f-f05c-4fb3-9f56-4bd83240ed81 Error: Runtime exited without providing a reason
Runtime.ExitError
```

Maybe some of you have worked with Lambda connected to ECS (Lambda Containe Image) and know what I am making bad?

**Reply by vmal** [reply]
Hi, were you able to figure out?
